src/server/campaignd/server.cpp

```diff
@@ -668,29 +753,111 @@ void server::handle_request_campaign_list(const server::request& req)
 
 void server::handle_request_campaign(const server::request& req)
 {
-	config& campaign = get_campaign(req.cfg["name"]);
+	config& campaign = get_addon(req.cfg["name"]);
 
 	if(!campaign || campaign["hidden"].to_bool()) {
 		send_error("Add-on '" + req.cfg["name"].str() + "' not found.", req.sock);
 		return;
 	}
-	
-	const int size = filesystem::file_size(campaign["filename"]);
 
-	if(size < 0) {
-		send_error("Add-on '" + req.cfg["name"].str() + "' could not be read by the server.", req.sock);
-		return;
+	// The desired version is selected on the client side considering
+	// the min_wesnoth_version in [version] children of the addon info (#TODO: gfgtdf's part)
+	std::string to = req.cfg["version"].str();
+	const std::string& from = req.cfg["from_version"].str();
+	std::string full_pack = campaign["filename"].str();
+
+	auto version_map = get_version_map(campaign);
+
+	if(version_map.empty()) {
+		// Old format ?
+		ERR_CS << "The (" + to + ") version of the addon '" << req.cfg["name"].str()
+			   << "' is stored in the old format (bug?)! Trying to send a legacy full-pack.";
+		to = campaign["version"].str();
+	} else {
+		auto version = version_map.find(version_info(to));
+		if(version != version_map.end()) {
+			full_pack += version->second["filename"].str();
+		} else {
+			WRN_CS << "The selected version (" << to << ") for the addon '" << req.cfg["name"].str()
+				   << "' has not been found or it is unspecified! Sending the latest version instead!\n";
+			to = version_map.rbegin()->first;
+			full_pack += version_map.rbegin()->second["filename"].str();
+		}
+
+		// Negotiate an update pack if possible
+		if(!from.empty() && version_map.count(version_info(from)) != 0) {
```

Comment:
If `version_map` is a `std::map`, the result of `count()` can only be 1 or 0. Please use `!empty()` for clarity.